drivers/espi/espi_taf_npcx.c

```diff
@@ -0,0 +1,466 @@
+	const struct espi_saf_pr *preg = pr->pregions;
+	size_t n = pr->nregions;
+	uint8_t regnum;
+	uint16_t bitmask, offset;
+	uint32_t rw_pr, override_rw;
+
+	if ((dev == NULL) || (pr == NULL)) {
+		return -EINVAL;
+	}
+
+	if (pr->nregions >= CONFIG_ESPI_TAF_PR_NUM) {
+		return -EINVAL;
+	}
+
+	while (n--) {
+		regnum = preg->pr_num;
+
+		if (regnum >= CONFIG_ESPI_TAF_PR_NUM) {
+			return -EINVAL;
+		}
+
+		rw_pr = preg->master_bm_we << NPCX_FRGN_WPR;
+		rw_pr = rw_pr | (preg->master_bm_rd << NPCX_FRGN_RPR);
+
+		if (preg->flags) {
+			bitmask = BIT_MASK(GET_FIELD_SZ(NPCX_FLASH_PRTR_BADDR));
+			offset = GET_FIELD_POS(NPCX_FLASH_PRTR_BADDR);
+			inst->FLASH_PRTR_BADDR[regnum] = ((preg->start & bitmask) << offset)
+							 | rw_pr;
+			bitmask = BIT_MASK(GET_FIELD_SZ(NPCX_FLASH_PRTR_HADDR));
+			offset = GET_FIELD_POS(NPCX_FLASH_PRTR_HADDR);
+			inst->FLASH_PRTR_HADDR[regnum] = (preg->end & bitmask) << offset;
+		}
+
+		override_rw = (preg->override_r << 16) | preg->override_w;
+		inst->FLASH_RGN_TAG_OVR[regnum] = override_rw;
+		preg++;
+	}
+
+	return 0;
+}
+
+static int espi_taf_npcx_activate(const struct device *dev)
+{
+	struct espi_reg *const inst = HAL_INSTANCE(dev);
+
+	inst->FLASHCTL &= ~BIT(NPCX_FLASHCTL_AUTO_RD_DIS_CTL);
+	inst->FLASHCTL &= ~BIT(NPCX_FLASHCTL_BLK_FLASH_NP_FREE);
+
+	return 0;
+}
+
+static bool espi_taf_npcx_channel_ready(const struct device *dev)
+{
+	struct espi_reg *const inst = HAL_INSTANCE(dev);
+
+	if (!IS_BIT_SET(inst->ESPICFG, NPCX_ESPICFG_FLCHANMODE)) {
+		return false;
+	}
+	return true;
+}
+
+/* This routine set FLASH_C_AVAIL for standard request */
+static void taf_set_flash_c_avail(const struct device *dev)
+{
+	struct espi_reg *const inst = HAL_INSTANCE(dev);
+	uint32_t tmp = inst->FLASHCTL;
+
+	/*
+	 * Clear FLASHCTL_FLASH_NP_FREE to avoid host puts a flash
+	 * standard request command at here.
+	 */
+	tmp &= NPCX_FLASHCTL_ACCESS_MASK;
+
+	/* Set FLASHCTL_FLASH_TX_AVAIL */
+	tmp |= BIT(NPCX_FLASHCTL_FLASH_TX_AVAIL);
+	inst->FLASHCTL = tmp;
+}
+
+/* This routine release FLASH_NP_FREE for standard request */
+static void taf_release_flash_np_free(const struct device *dev)
+{
+	struct espi_reg *const inst = HAL_INSTANCE(dev);
+	uint32_t tmp = inst->FLASHCTL;
+
+	/*
+	 * Clear FLASHCTL_FLASH_TX_AVAIL to avoid host puts a
+	 * GET_FLASH_C command at here.
+	 */
+	tmp &= NPCX_FLASHCTL_ACCESS_MASK;
+
+	/* Release FLASH_NP_FREE */
+	tmp |= BIT(NPCX_FLASHCTL_FLASH_NP_FREE);
+	inst->FLASHCTL = tmp;
+}
+
+static int taf_npcx_completion_handler(const struct device *dev, uint32_t *buffer)
+{
+	uint16_t size = DIV_ROUND_UP((uint8_t)(buffer[0]) + 1, sizeof(uint32_t));
+	struct espi_reg *const inst = HAL_INSTANCE(dev);
+	struct npcx_taf_head *head = (struct npcx_taf_head *)buffer;
+	uint8_t i;
+
+	/* Check the Flash Access TX Queue is empty by polling
+	 * FLASH_TX_AVAIL.
+	 */
+	if (WAIT_FOR(IS_BIT_SET(inst->FLASHCTL, NPCX_FLASHCTL_FLASH_TX_AVAIL),
+		      10000, k_busy_wait(1))) {
```

Comment:
If you just want to spin you can drop the last argument and just use NULL, the macro spins with `Z_SPIN_DELAY(10)` for you

Also I suggest using a define for `10000`, if anything to add a time suffix

Reply:
Done.